### PR TITLE
Move static/settings.js to TypeScript

### PR DESCRIPTION
**Pull request by RubenRBS**
This translates the settings file to TS

### CODE REVIEW

File: static/settings.ts [hunk 1]
```diff
@@ -0,0 +1,293 @@
...
+// ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+// LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+// CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+// SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+// INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+// CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+// ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+// POSSIBILITY OF SUCH DAMAGE.
+
+import $ from 'jquery';
```
Comment (matsjla) [inline static/settings.ts:25]: jQuery is globally injected so we can drop this import

File: static/settings.ts [hunk 2]
```diff
@@ -0,0 +1,293 @@
...
+const themes = require('./themes').themes
+
+class ISetting {
+    elem: JQuery;
+    name: string;
+
+    constructor(elem, name) {
+        this.elem = elem;
+        this.name = name;
+    }
```
Comment (matsjla) [inline static/settings.ts:41]: Can use the data object shorthand

```suggestion
    constructor(public elem: JQuery, public name: string) {}
```

File: static/settings.ts [hunk 3]
```diff
@@ -0,0 +1,293 @@
...
+        return !!this.elem.prop('checked');
+    }
+
+    override putUi(value) {
+        this.elem.prop('checked', !!value);
+    }
+}
+
+class Select extends ISetting {
+    constructor(elem, name, populate) {
```
Comment (matsjla) [inline static/settings.ts:63]: Can add types to the constructor arguments

```suggestion
    // TODO: add a better type for populate argument
    constructor(elem: JQuery, name: string, populate: any[]) {
```

File: static/settings.ts [hunk 4]
```diff
@@ -0,0 +1,293 @@
...
+        }
+    }
+
+    override putUi(value) {
+        this.elem.val(value.toString());
+    }
+}
+
+class Slider extends ISetting {
+    constructor(elem, name, sliderSettings) {
```
Comment (matsjla) [inline static/settings.ts:78]: Could type arguments as

```suggestion
    constructor(elem: JQuery, name: string, sliderSettings: Record<'max' | 'min' | 'step', number>) {
```

File: static/settings.ts [hunk 5]
```diff
@@ -0,0 +1,293 @@
...
+    }
+}
+
+class Textbox extends ISetting {}
+
+class Numeric extends ISetting {
+    private readonly min: number;
+    private readonly max: number;
+
+    constructor(elem, name, params) {
```
Comment (matsjla) [inline static/settings.ts:98]: Could type arguments as

```suggestion
    constructor(elem: JQuery, name: string, params: Record<'min' | 'max', number>) {
```

File: static/settings.ts [hunk 6]
```diff
@@ -0,0 +1,293 @@
...
+
+// Ignore max statements, there's no limit as to how many settings we need
+// eslint-disable-next-line max-statements
+export function setupSettings(root: JQuery, settings, onChange: (SiteSettings) => void, subLangId: string | null) {
+    settings = settings || {};
+    // Ensure the default language is not "null" but undefined. Temporary patch for a previous bug :(
+    settings.defaultLanguage = settings.defaultLanguage === null ? undefined : settings.defaultLanguage;
+    const settingsObjs: ISetting[] = [];
+
+    var currentSettings = settings;
```
Comment (matsjla) [inline static/settings.ts:129]: Can replace all instances of var with let/const

**Review by matsjla — CHANGES_REQUESTED** [review]
Looks good overall! I've suggested a couple typescript types that could be added and some other shortcuts, but I'm also happy with it as-is.

File: static/settings.ts [hunk 7]
```diff
@@ -0,0 +1,287 @@
...
+import $ from 'jquery';
+
+import { SiteSettings } from './settings.interfaces';
+import { options } from './options';
+import * as colour from './colour';
+import * as local from './local';
+
+const themes = require('./themes').themes
+
+class ISetting {
```
[the remark this answers is not in the document]
Comment (matsjla) [reply static/settings.ts:34]: BaseSetting might work

File: static/settings.ts [hunk 8]
```diff
@@ -0,0 +1,386 @@
...
+
+    private add<Type extends BaseSetting>(setting: Type, defaultValue: any) {
+        const key = setting.name;
+        if (this.settings[key] === undefined) this.settings[key] = defaultValue;
+        this.settingsObjs.push(setting);
+        setting.elem.on('change', this.onUiChange.bind(this))
+    }
+
+    private addCheckboxes() {
+        const checkboxes: [string, keyof SiteSettings, boolean][]= [
```
Comment (matsjla) [inline static/settings.ts:239]: Could be nice to have a comment here saying what each array member is, maybe
```suggestion
        // Known checkbox options in order [selector, key, defaultValue]
        const checkboxes: [string, keyof SiteSettings, boolean][]= [
```

**Review by matsjla — APPROVED** [review]

**Comment by partouf** [comment]
this is now live

**Pull request merged**